From 9379a3ce95507d894e2fbbf452bbc49b951db400 Mon Sep 17 00:00:00 2001
From: John Linn <john.linn@xilinx.com>
Date: Thu, 22 Mar 2012 20:45:40 -0700
Subject: [PATCH] Xilinx: ARM: I2C: xilinx-ps: Fix hold flag handling

Set the hold flag whenever there is more data to send and clear it when sending
the last message. Otherwise STOP-START condintions will be generated instead of
a RESTART condition when sending multiple messages at once.

This was a patch from Lars that wouldn't apply so I manually added the change.

Signed-off-by: Lars-Peter Clausen <lars@metafoo.de>
---
 drivers/i2c/busses/i2c-xilinx_ps.c |   20 ++++++++++++--------
 1 files changed, 12 insertions(+), 8 deletions(-)

diff --git a/drivers/i2c/busses/i2c-xilinx_ps.c b/drivers/i2c/busses/i2c-xilinx_ps.c
index a5f44d1..71923eb 100644
--- a/drivers/i2c/busses/i2c-xilinx_ps.c
+++ b/drivers/i2c/busses/i2c-xilinx_ps.c
@@ -285,8 +285,10 @@ static void xi2cps_mrecv(struct xi2cps *id)
 	ctrl_reg = xi2cps_readreg(XI2CPS_CR_OFFSET);
 	ctrl_reg |= (XI2CPS_CR_RW_MASK | XI2CPS_CR_CLR_FIFO_MASK);
 
-	if (id->recv_count > XI2CPS_FIFO_DEPTH)
-		ctrl_reg |= XI2CPS_CR_HOLD_BUS_MASK;
+	if (id->recv_count > XI2CPS_FIFO_DEPTH || id->bus_hold_flag)
+		ctrl_reg |= XI2CPS_CR_HOLD_BUS_MASK;
+	else
+		ctrl_reg &= ~XI2CPS_CR_HOLD_BUS_MASK;
 
 	xi2cps_writereg(ctrl_reg, XI2CPS_CR_OFFSET);
 	xi2cps_writereg((id->p_msg->addr & XI2CPS_ADDR_MASK),
@@ -331,8 +333,10 @@ static void xi2cps_msend(struct xi2cps *id)
 	ctrl_reg &= ~XI2CPS_CR_RW_MASK;
 	ctrl_reg |= XI2CPS_CR_CLR_FIFO_MASK;
 
-	if ((id->send_count) > XI2CPS_FIFO_DEPTH)
-		ctrl_reg |= XI2CPS_CR_HOLD_BUS_MASK;
+	if ((id->send_count) > XI2CPS_FIFO_DEPTH || id->bus_hold_flag)
+		ctrl_reg |= XI2CPS_CR_HOLD_BUS_MASK;
+	else
+		ctrl_reg &= ~XI2CPS_CR_HOLD_BUS_MASK;
 	xi2cps_writereg(ctrl_reg, XI2CPS_CR_OFFSET);
 
 	/*
@@ -398,13 +402,13 @@ static int xi2cps_master_xfer(struct i2c_adapter *adap, struct i2c_msg *msgs,
 	}
 
 	/*
-	 * Set the flag to zero when multiple messages are to be
+	 * Set the flag to one when multiple messages are to be
 	 * processed with a repeated start.
 	 */
 	if (num > 1)
-		id->bus_hold_flag = 0;
-	else
-		id->bus_hold_flag = 1;
+		id->bus_hold_flag = 1;
+	else
+		id->bus_hold_flag = 0;
 
 	/* Process the msg one by one */
 	for (count = 0; count < num; count++, msgs++) {
-- 
1.7.0.4

